### DISCUSSION TITLE
How to retrieve the value from multiple json data

### DISCUSSION THREAD

**Discussion by max420max** [grafana/grafana-infinity-datasource#741] [Q&A]
Hi , I am using Infinite Plugin for Grafana , i have JSON file as below , I tried two approach, 1. UQL query and 2. adding the columns.
How to retrieve the value from multiple array?

```json
{
  "Results": [
    {
      "Name": "device-bff-pdb",
      "Namespace": "default-release-name-deviceportal",
      "Kind": "PodDisruptionBudget",
      "Results": {
        "pdbDisruptionsIsZero": {
          "ID": "pdbDisruptionsIsZero",
          "Message": "Voluntary evictions are possible",
          "Details": null,
          "Success": true,
          "Severity": "warning",
          "Category": "Reliability",
          "Mutations": null
        }
      },
      "PodResult": null,
      "CreatedTime": "0001-01-01T00:00:00Z"
    },
    {
      "Name": "ca-certificates-configmap",
      "Namespace": "default-release-name-aops",
      "Kind": "ConfigMap",
      "Results": {
        "sensitiveConfigmapContent": {
          "ID": "sensitiveConfigmapContent",
          "Message": "The ConfigMap does not contain potentially sensitive content in its keys and values",
          "Details": null,
          "Success": true,
          "Severity": "danger",
          "Category": "Security",
          "Mutations": null
        }
      },
      "PodResult": null,
      "CreatedTime": "0001-01-01T00:00:00Z"
    }
  ]
}
```

======================

Results.Results.**pdbDisruptionsIsZero**.Message

Results.Results.**sensitiveConfigmapContent**.Message

how to retrive those two value ?

**Comment by yesoreyeram** [comment]
You can use the backend parser and `Results.*.*` (jsonata selector) as your root selector.

<img width="1683" alt="image" src="https://github.com/grafana/grafana-infinity-datasource/assets/153843/85ed536b-2314-4a79-8da9-7c8eb78d8af1">

**Reply by max420max** [reply]
Thank you very much for reply.



Kind , Namespace , Name column data is not retrived. kindly help

**Reply by yesoreyeram** [reply]
In the rows/Root selector, you can write a proper [jsonata selector](https://docs.jsonata.org/overview.html) instead of just `Results`

**Reply by max420max** [reply]
I am new in JSON query , not sure where is wrong . i tried two different query but not getting expected output.

Results.{
  Name: Name,
  Namespace: Namespace,
  Kind: Kind,
  Message: Results.pdbDisruptionsIsZero.Message & Results.sensitiveConfigmapContent.Message,
  Severity: Results.pdbDisruptionsIsZero.Severity &  Results.sensitiveConfigmapContent.Severity,
  Category: Results.pdbDisruptionsIsZero.Category &  Results.sensitiveConfigmapContent.Category,
  Success: Results.pdbDisruptionsIsZero.Success &  Results.sensitiveConfigmapContent.Success
}

Results.{
  "Name": Name,
  "Namespace": Namespace,
  "Kind": Kind,
  "Results": Results.*{
    "ID": ID,
    "Message": Message,
    "Success": Success,
    "Severity": Severity,
    "Category": Category
  }
}

The same query gives the correct output in https://try.jsonata.org/ but when same query i am trying to use in Grafana Row/Root it does not give output.

However Results.*.* gives the result but not fetching the data of three columns(Kind , Namespace , Name)

**Reply by yesoreyeram** [reply]
Can you share the query you tried which was working jsonata?

**Reply by max420max** [reply]
Query shared in below thread !

**Reply by max420max** [reply]
[json-data.json](https://github.com/grafana/grafana-infinity-datasource/files/13955770/json-data.json)
For reference, i am attaching the complete JSON data,
The below query gives me the CORRECT result except for the few arrays (automountServiceAccountToken, cpuLimitsMissing)''

"ContainerResults": [
{
"Name": "macma-configuration-operator-pod",
"Results": {
"cpuLimitsMissing": {
"ID": "cpuLimitsMissing",
"Message": "CPU limits are set",
"Details": null,
"Success": true,
"Severity": "warning",
"Category": "Efficiency",
"Mutations": null
},
"PodResult": {
"Name": "",
"Results": {
"automountServiceAccountToken": {
"ID": "automountServiceAccountToken",
"Message": "The ServiceAccount will be automounted",
"Details": null,
"Success": false,
"Severity": "warning",
"Category": "Security",
"Mutations": null
},

How to retrive those two array data? I tried to adding those two array ( cpuLimitsMissing , automountServiceAccountToken) in query but it does not work.

Below query too long , how to improve here ?

parse-json
| project "Results"
| project-away "CreatedTime"
| extend "Kind", "Name","Namespace", "ID"=strcat("Results.pdbDisruptionsIsZero.ID","Results.sensitiveConfigmapContent.ID","Results.rolebindingClusterAdminClusterRole.ID","Results.rolebindingClusterAdminRole.ID","Results.rolebindingClusterRolePodExecAttach.ID","Results.rolebindingRolePodExecAttach.ID","Results.rolePodExecAttach.ID"),"Message"=strcat("Results.pdbDisruptionsIsZero.Message","Results.sensitiveConfigmapContent.Message","Results.rolebindingClusterAdminClusterRole.Message","Results.rolebindingClusterAdminRole.Message","Results.rolebindingClusterRolePodExecAttach.Message","Results.rolebindingRolePodExecAttach.Message","Results.rolePodExecAttach.Message"), "Severity"=strcat("Results.pdbDisruptionsIsZero.Severity","Results.sensitiveConfigmapContent.Severity", "Results.rolePodExecAttach.Severity"),"Category"=strcat("Results.pdbDisruptionsIsZero.Category","Results.sensitiveConfigmapContent.Category","Results.rolebindingClusterAdminClusterRole.Category","Results.rolebindingClusterAdminRole.Category","Results.rolebindingClusterRolePodExecAttach.Category","Results.rolebindingRolePodExecAttach.Category","Results.rolePodExecAttach.Category"), "Success"=strcat("Results.pdbDisruptionsIsZero.Success","Results.sensitiveConfigmapContent.Success","Results.rolebindingClusterAdminClusterRole.Success","Results.rolebindingClusterAdminRole.Success","Results.rolebindingClusterRolePodExecAttach.Success","Results.rolebindingRolePodExecAttach.Success","Results.rolePodExecAttach.Success")
| project-away "PodResult", "Results"

**Comment by max420max** [comment]
[json-data.json](https://github.com/grafana/grafana-infinity-datasource/files/13955770/json-data.json)
For reference, i am attaching the complete JSON data,
The below query gives me the CORRECT result except for the few arrays (automountServiceAccountToken, cpuLimitsMissing)''

"ContainerResults": [
          {
            "Name": "macma-configuration-operator-pod",
            "Results": {
              "cpuLimitsMissing": {
                "ID": "cpuLimitsMissing",
                "Message": "CPU limits are set",
                "Details": null,
                "Success": true,
                "Severity": "warning",
                "Category": "Efficiency",
                "Mutations": null
              },
"PodResult": {
        "Name": "",
        "Results": {
          "automountServiceAccountToken": {
            "ID": "automountServiceAccountToken",
            "Message": "The ServiceAccount will be automounted",
            "Details": null,
            "Success": false,
            "Severity": "warning",
            "Category": "Security",
            "Mutations": null
          },

How to retrive those two array data? I tried to adding those two array ( cpuLimitsMissing , automountServiceAccountToken) in query but it does not work.

Below query too long , how to improve here ?

parse-json
| project "Results"
| project-away "CreatedTime"
| extend "Kind", "Name","Namespace", "ID"=strcat("Results.pdbDisruptionsIsZero.ID","Results.sensitiveConfigmapContent.ID","Results.rolebindingClusterAdminClusterRole.ID","Results.rolebindingClusterAdminRole.ID","Results.rolebindingClusterRolePodExecAttach.ID","Results.rolebindingRolePodExecAttach.ID","Results.rolePodExecAttach.ID"),"Message"=strcat("Results.pdbDisruptionsIsZero.Message","Results.sensitiveConfigmapContent.Message","Results.rolebindingClusterAdminClusterRole.Message","Results.rolebindingClusterAdminRole.Message","Results.rolebindingClusterRolePodExecAttach.Message","Results.rolebindingRolePodExecAttach.Message","Results.rolePodExecAttach.Message"), "Severity"=strcat("Results.pdbDisruptionsIsZero.Severity","Results.sensitiveConfigmapContent.Severity", "Results.rolePodExecAttach.Severity"),"Category"=strcat("Results.pdbDisruptionsIsZero.Category","Results.sensitiveConfigmapContent.Category","Results.rolebindingClusterAdminClusterRole.Category","Results.rolebindingClusterAdminRole.Category","Results.rolebindingClusterRolePodExecAttach.Category","Results.rolebindingRolePodExecAttach.Category","Results.rolePodExecAttach.Category"), "Success"=strcat("Results.pdbDisruptionsIsZero.Success","Results.sensitiveConfigmapContent.Success","Results.rolebindingClusterAdminClusterRole.Success","Results.rolebindingClusterAdminRole.Success","Results.rolebindingClusterRolePodExecAttach.Success","Results.rolebindingRolePodExecAttach.Success","Results.rolePodExecAttach.Success")
| project-away "PodResult", "Results"
